internal/dispatch/remote/cluster.go

```diff
@@ -75,18 +89,99 @@ func (cr *clusterDispatcher) DispatchCheck(ctx context.Context, req *v1.Dispatch
 
 	ctx = context.WithValue(ctx, consistent.CtxKey, requestKey)
 
-	withTimeout, cancelFn := context.WithTimeout(ctx, cr.dispatchOverallTimeout)
-	defer cancelFn()
+	resp, err := dispatchRequest(ctx, cr, "check", req, func(ctx context.Context, client ClusterClient) (*v1.DispatchCheckResponse, error) {
+		resp, err := client.DispatchCheck(ctx, req)
+		if err != nil {
+			return resp, err
+		}
 
-	resp, err := cr.clusterClient.DispatchCheck(withTimeout, req)
+		err = adjustMetadataForDispatch(resp.Metadata)
+		return resp, err
+	})
 	if err != nil {
 		return &v1.DispatchCheckResponse{Metadata: requestFailureMetadata}, err
 	}
 
-	err = adjustMetadataForDispatch(resp.Metadata)
 	return resp, err
 }
 
+type respTuple[S any] struct {
+	resp *S
+	err  error
+}
+
+func dispatchRequest[Q zerolog.LogObjectMarshaler, S any](ctx context.Context, cr *clusterDispatcher, reqKey string, req Q, handler func(context.Context, ClusterClient) (*S, error)) (*S, error) {
+	withTimeout, cancelFn := context.WithTimeout(ctx, cr.dispatchOverallTimeout)
+	defer cancelFn()
+
+	if len(cr.secondaryDispatchExprs) == 0 || len(cr.secondaryDispatch) == 0 {
+		return handler(withTimeout, cr.clusterClient)
+	}
+
+	// If no secondary dispatches are defined, just invoke directly.
+	expr, ok := cr.secondaryDispatchExprs[reqKey]
+	if !ok {
+		return handler(withTimeout, cr.clusterClient)
+	}
+
+	// Otherwise invoke in parallel with any secondary matches.
+	primaryResultChan := make(chan respTuple[S], 1)
+	secondaryResultChan := make(chan *S, len(cr.secondaryDispatch))
+
+	// Run the main dispatch.
+	go func() {
+		resp, err := handler(withTimeout, cr.clusterClient)
+		primaryResultChan <- respTuple[S]{resp, err}
+	}()
+
+	result, err := RunDispatchExpr(expr, req)
+	if err != nil {
+		log.Warn().Err(err).Msg("error when trying to evaluate the dispatch expression")
+	}
+
+	log.Trace().Str("secondary-dispatchers", strings.Join(result, ",")).Object("request", req).Msg("running secondary dispatchers")
+	for _, secondary := range cr.secondaryDispatch {
+		secondary := secondary
+
+		if !stringz.SliceContains(result, secondary.Name) {
+			continue
+		}
+
+		log.Trace().Str("secondary-dispatcher", secondary.Name).Object("request", req).Msg("running secondary dispatcher")
+		go func() {
+			resp, err := handler(withTimeout, secondary.Client)
+			if err != nil {
+				// For secondary dispatches, ignore any errors, as only the primary will be handled in
+				// that scenario.
+				log.Trace().Str("secondary", secondary.Name).Err(err).Msg("got ignored secondary dispatch error")
+				return
+			}
+
+			secondaryResultChan <- resp
+		}()
+	}
+
+	var foundError error
+	select {
+	case <-withTimeout.Done():
+		return nil, fmt.Errorf("check dispatch has timed out")
+
+	case r := <-primaryResultChan:
```

Comment:
it would be nice to have a metric for secondary/primary dispatches. A counter with a label for which dispatcher handled it?

Reply:
Added